adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,712 @@
+}
+
+static const IO_INTERFACE_DESCRIPTION tlsio_openssl_interface_description =
+{
+    tlsio_openssl_retrieveoptions,
+    tlsio_openssl_create,
+    tlsio_openssl_destroy,
+    tlsio_openssl_open,
+    tlsio_openssl_close,
+    tlsio_openssl_send,
+    tlsio_openssl_dowork,
+    tlsio_openssl_setoption
+};
+
+static LOCK_HANDLE * openssl_locks = NULL;
+
+static struct CRYPTO_dynlock_value* openssl_dynamic_locks_create_cb(const char* file, int line)
+{
+    struct CRYPTO_dynlock_value* result;
+    return result;
+}
+
+static void openssl_dynamic_locks_uninstall(void)
+{
+}
+
+static void openssl_dynamic_locks_install(void)
+{
+}
+
+static void openssl_static_locks_lock_unlock_cb(int lock_mode, int lock_index, const char * file, int line)
+{
+}
+
+static void indicate_open_complete(TLS_IO_INSTANCE* tls_io_instance, IO_OPEN_RESULT open_result)
+{
+    if (tls_io_instance->on_io_open_complete == NULL)
+    {
+        LogError("NULL on_io_open_complete.");
+    }
+    else
+    {
+        tls_io_instance->on_io_open_complete(tls_io_instance->on_io_open_complete_context, open_result);
+    }
+}
+
+
+static int lwip_net_errno(int fd)
+{
+    int sock_errno = 0;
+    u32_t optlen = sizeof(sock_errno);
+    getsockopt(fd, SOL_SOCKET, SO_ERROR, &sock_errno, &optlen);
+    return sock_errno;
+}
+
+static void lwip_set_non_block(int fd) 
+{
+  int flags = -1;
+  int error = 0;
+
+  while(1){
+      flags = fcntl(fd, F_GETFL, 0);
+      if (flags == -1){
+          error = lwip_net_errno(fd);
+          if (error != EINTR){
+              break;
+          }
+      } else{
+          break;
+      }
+  }
+
+  while(1){
+      flags = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+      if (flags == -1) {
+          error = lwip_net_errno(fd);
+          if (error != EINTR){
+              break;
+          }
+      } else{
+          break;
+      }
+  }
+
+}
+
+
+LOCAL void openssl_thread_LWIP_CONNECTION(void *p)
+{
+    int ret = -1;
+
+    int sock = -1;
+    struct sockaddr_in sock_addr;
+    fd_set readset;
+    fd_set writeset;
+    fd_set errset;
+
+    ip_addr_t target_ip;
+    SSL_CTX *ctx;
+    SSL *ssl;
+
+    TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)p;
+
+    LogInfo("OpenSSL thread start...\n");
+
+    do {
+        ret = netconn_gethostbyname(tls_io_instance->hostname, &target_ip);
+    } while(ret);
+    // This is useful for debugging purpose.
+    // LogInfo("get target IP is %d.%d.%d.%d\n", (unsigned char)((target_ip.addr & 0x000000ff) >> 0),
+    //                                             (unsigned char)((target_ip.addr & 0x0000ff00) >> 8),
+    //                                             (unsigned char)((target_ip.addr & 0x00ff0000) >> 16),
+    //                                             (unsigned char)((target_ip.addr & 0xff000000) >> 24));
+
+    LogInfo("create socket ......");
+    sock = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock < 0) {
+        LogError("create socket failed\n");
+        return;
```

Comment:
Our code guidelines do not allow returns in the middle of the function. Please use only one return at the end of the function.